modules/SequencesExt.tla

```diff
@@ -124,6 +124,53 @@ Last(s) ==
 
 -----------------------------------------------------------------------------
 
+SubSeqs(s) ==
+  (**************************************************************************)
+  (* The set sequence of all subsequences of the sequence  s  .  Note that  *)
+  (* the empty sequence  <<>>  is defined to be a subsequence of any        *)
+  (* sequence.                                                              *)
+  (**************************************************************************)
+  { SubSeq(s, i+1, j) : i, j \in 0..Len(s) }
+
+IndexFirstSubSeq(s, t) ==
+  (**************************************************************************)
+  (**************************************************************************)
+  LET last == CHOOSE i \in 0..Len(t) :
+                /\ s \in SubSeqs(SubSeq(t, 1, i))
+                /\ \A j \in 0..i-1 : s \notin SubSeqs(SubSeq(t, 1, j))
+  IN last - (Len(s) - 1)
+
+ReplaceSubSeqAt(i, r, s, t) ==
+  (**************************************************************************)
+  (* The sequence  t  with its subsequence  s  at position  i  replaced by  *)
+  (* the sequence  r  .                                                     *)
+  (**************************************************************************)
+  LET prefix == SubSeq(t, 1, i - 1)
+      suffix == SubSeq(t, i + Len(s), Len(t))
+  IN prefix \o r \o suffix 
+
+ReplaceFirstSubSeq(r, s, t) ==
+  (**************************************************************************)
+  (* The sequence  t  with its subsequence  s  replaced by the sequence  r  *)
+  (**************************************************************************)
+  IF s \in SubSeqs(t)
+  THEN ReplaceSubSeqAt(IndexFirstSubSeq(s, t), r, s, t)
+  ELSE t
+
+RECURSIVE ReplaceAllSubSeqs(_,_,_)
+ReplaceAllSubSeqs(r, s, t) ==
```

Comment:
Looks good to me, but I believe that

ReplaceAllSubSeqs(<<1,1>>, <<1>>, <<1,1>>)

loops, this is probably what is expected? Perhaps clarify in a comment. 
Alternatively, one would have to keep track of the starting index for replacement in the recursive call in order to obtain

ReplaceAllSubSeqs(<<1,1>>, <<1>>, <<1,1>>) = <<1,1,1,1>>